specification/dartLangSpec.tex

```diff
@@ -17613,31 +17607,87 @@ \subsection{Type Aliases}
 (\ref{overview}).%
 }
 
+\LMHash{}%
+We define the notion that a type alias declaration
+\NoIndex{expands to a type variable}
+inductively as follows:
+When $j \in 1 .. s$, we say that a type alias declaration of the form
+\code{\TYPEDEF{} $F$<\TypeParametersStd> = $X_j$;}
+\NoIndex{expands to a type variable at position} $j$.
+Similarly, if $H$ denotes a type alias declaration
+that declares $k$ type parameters and
+expands to a type variable at position $i \in 1 .. k$,
+we say that the type alias declaration
+
+\noindent
+\code{\TYPEDEF{} $G$<\TypeParametersStd> = $H$<\List{T}{1}{s}>;}
+
+\noindent
+\NoIndex{expands to a type variable at position} $j$
+when $T_i$ is $X_j$ for some $j \in 1 .. s$.
+We say that a type alias declaration
+\IndexCustom{expands to a type variable}{%
+  type alias!expands to a type variable}
+if it expands to a type variable at any position.
+
 \commentary{%
-A type alias can be used as a type annotation,
-as a return type or parameter type in a function declaration,
-in a function type,
-as a type argument,
-in a type test,
-in a type cast,
-and in an \ON{} clause of a \TRY{} statement.
-
-Consider the case where the body of a given type alias $F$
-is a \synt{typeName} that denotes a non-generic class,
-or it is a parameterized type that starts with
-a \synt{typeName} that denotes a generic class,
-or one of these cases occur indirectly via another type alias.
-In this case $F$ or a parameterized type that starts with $F$,
-whichever is not an error,
-can also be used to name a constructor in an instance creation expression
-(\ref{instanceCreation}),
-and it can be used as a superclass, mixin, or superinterface
-(\ref{superclasses}, \ref{superinterfaces}, \ref{mixinApplication}).
-Moreover, $F$ or a parameterized type that starts with $F$,
-whichever is not an error,
-can be used to invoke static methods of the denoted class.
+This means that if a type alias declaration named $F$
+expands to a type variable
+then \code{$F$<\List{T}{1}{s}>} is $T_j$ for some $j \in 1 .. s$.
+If the declaration does not expand to a type variable then that never occurs.%
 }
 
+\LMHash{}%
+Let $T$ be a parameterized type \code{$F$<$T_1, \ldots T_k$>}
+in a scope where $F$ denotes a type alias declaration $D$.
+We say that $T$
+\IndexCustom{uses $D$ as a class}{type alias!used as a class}
+when $T$ occurs in one of the following ways:
+
+\begin{itemize}
+\item $T$ denotes a constructor in an instance creation expression
+  (\ref{instanceCreation}).
+\item $T$ is the redirectee in a redirecting factory constructor
+  (\ref{redirectingFactoryConstructors}).
+\item $T$ is used as a superclass, a mixin, or a superinterface
+  in a class declaration
+  (\ref{superclasses}, \ref{mixinApplication}, \ref{superinterfaces}).
+\end{itemize}
+
+\LMHash{}%
+A compile-time error occurs if $D$ expands to a type variable
+and $T$ uses $D$ as a class.
```

Comment:
There are several situations where we could allow a type alias that expands to a type argument to be used as a class. The main issue has been that an instance creation is an expression, and it is subject to type inference. This would introduce the notion that type inference should select a constructor.

Consider `Iterable<num> iter = F.id();` where `typedef F<X> = X;`, we'd then expect the type inference to search among all classes in order to find one, `C`, (and perhaps some type arguments for it, `T1,.. Tk`), such that `C.id` (respectively `C<T1, .. Tk>.id`) denotes a constructor.

That's a highly irregular (and probably error prone) feature.

But it is probably a similarly deep can of worms if we allow a type argument of a class to determine a superclass (like your `C<T> extends Id<T>` example).

We could allow `class D extends Id<C> {}`, but I don't think it's worth the trouble to start outlining precisely in which situations it is an error to use a type alias "as a class", I think we should simply block that road as being too dangerous and not worth the trouble.

Reply:
Let's keep it as-is for now,